### PR TITLE
Constraints for generic tuple types

### LINKED ISSUE #53256: Variadic element at position does not match element in target
# Bug Report

### 🔎 Search Terms 
Tuple spread variadic element length

### 🕗 Version & Regression Information

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230313&ssl=14&ssc=2&pln=1&pc=1#code/GYVwdgxgLglg9mABAQ0QHgNKIKYA8rZgAmAzogNrJgCeAugHyIAUAlIgN4CwAUIogPT9EAASgkAtDADmYOACdsiALJwiMYDGxyyEKrKgoADoezI5iABZbsAOiYBGewA4ALCx58i2CABsziiAQSAwh7AC4KGyiMWh4PASFRCWlZBWVVdU1tRF0wfSMTM0trO0dXd15EL19-HKCQgCYI8iibDAAaRFaYuMrBRAB5AGt4wLBgxAAjcIoqOkQAXhz7Xr5+gBVqE0QAchbozu7aHcQYMnzkEhIU5EmfRSg4RCgtxT25zrnjm3j+vgA1MwwZBqCA4e4AW0IBmQBkMcGusAQiAADKckCQ4CA5BBFEQ4NhznADBDYRALODsFCwDC4QiYEikGiYEgoGYpNgoHYGgBmBoNCp8MYTSZNWY0T40WiLHINHgAXyAA)

### 💻 Code

```ts
function a <K extends [any]> () {
  // @ts-ignore Modifiers cannot appear here.(1184)
  declare const c1: [...K]

  // @ts-ignore Modifiers cannot appear here.(1184)
  declare const c2: [...K, ...K]

  // Ok
  const b1: [any] = c1

  // Type '[...K, ...K]' is not assignable to type '[any, any]'.
  //   Variadic element at position 0 in source does not match element at position 0 in target.(2322)
  const b2: [any, any] = c2
}
```

### 🙁 Actual behavior

c2 is not assignable to b2

### 🙂 Expected behavior

c2 is assignable to b2

### LINKED ISSUE #53563: Type '[...T]["length"]' is not assignable to type 'string | number | bigint | boolean | null | undefined'
# Bug Report

### 🔎 Search Terms
Template Literal Types

### 🕗 Version & Regression Information
This is the behavior in every version I tried, and I reviewed the FAQ for entries about type alias preservation

### 💻 Code

```ts
type ToStringLength1<T extends any[]> = `${T['length']}` // ok
type ToStringLength2<T extends any[]> = `${[...T]['length']}` // error 
```

### 🙁 Actual behavior

type conversion error

### 🙂 Expected behavior

no error

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #53256.
Fixes #53563.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by awerlogus** [comment]
@ahejlsberg will this PR fix also https://github.com/microsoft/TypeScript/issues/53256?

**Comment by ahejlsberg** [comment]
@awerlogus Yes, I'll make sure it does.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot test top100

**Comment by awerlogus** [comment]
@ahejlsberg please take a look at these two examples which are also related to generic tuple types

```ts
// No error, but expected
const a = <T extends [unknown]>(x: T): T[0] => 123

// string type, but number value
const b = a([''])
```

```ts
declare const a: <T extends [unknown]>(x: T[0]) => void

// const a: <[unknown]>(x: unknown) => void
a('123')

// It would be nice if type of 0 element of T will be inferred from type of argument passed as x
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13820,6 +13820,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (t.flags & TypeFlags.Substitution) {
                 return getBaseConstraint(getSubstitutionIntersection(t as SubstitutionType));
             }
+            if (isGenericTupleType(t)) {
+                // We substitute constrains for variadic elements only when the constrains are array or tuple types
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts:13824]:
```suggestion
                // We substitute constraints for variadic elements only when the constraints are array or tuple types

```

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
Tests all look good and performance is unaffected (though some noise in the various runs). This one is ready to merge.

**Review by jakebailey — APPROVED** [review]

**Comment by jcalz** [comment]
Is this a separate issue?

```ts
type NTuple<N extends number, A extends any[] = []> =
    A['length'] extends N ? A : NTuple<N, [any, ...A]>;

type X<N extends number> = `${[...NTuple<N>]['length']}`; // error
//                                           ^?
// Type '[...NTuple<N, []>]["length"]' is not assignable to type 
// 'string | number | bigint | boolean | null | undefined'.(2322)
```
[Playground link](https://www.typescriptlang.org/play?ts=5.1.0-dev.20230513#code/C4TwDgpgBAcgKgVzAGwgHhlCAPYEB2AJgM5T4IC2ARhAE4A0UAglrgSVAIb4gDaAulAC8UAQD5hAKCgzmvAOSp8Ac2AALeYJx4ipTAH5mUAFyxEKdDEa9uIRgDpHTfmIDckyaEhQAGhlY6HOTUdBIiAAYAJADevI728EioGGL8CkqqGvwAvuGuUAD0BVi0tAD2tJJFsjW1dfUNjVAAevpVxXDg0PJxjokWGNYuaQBEGeoj-PJQAJak+GXAXMTEM8r4nFSoUMBlO11Q7VDyxMC0MypQAD5klDS011BUaxdLN1RlZajcj+TIyI8EEQIAAzC4QQjyewACgATABmWGwgCUkiAA)

Seems related but maybe something about generic *recursive* tuple types makes this different?  Comes from [this SO question](https://stackoverflow.com/q/76243178/2887218)

**Pull request merged**